### PR TITLE
fix: Honour machine-type from config file above the one coming from metadata server

### PR DESCRIPTION
**Pull request by gargnitingoogle**
### Description

This pull request addresses a critical bug where the `machine-type` setting from the configuration file was not being respected, leading to unexpected behavior as it was being superseded by values from the metadata server. The changes introduce a robust precedence mechanism, ensuring that the `machine-type` is correctly identified first by CLI flags, then by the config file, and only then by the metadata server. This provides more predictable and controllable machine type identification for the application.
- There is a known-issue documented in #4087 for this issue.
- This issue started with #3799 .

#### Highlights

* **Bug Fix: Machine Type Precedence**: This pull request resolves an issue where the `machine-type` specified in the configuration file was being incorrectly overridden by values obtained from the metadata server. The fix ensures that the `machine-type` is now honored according to a defined precedence.
* **Precedence Order for Machine Type**: A clear precedence order has been established for determining the `machine-type`: command-line interface (CLI) flags take highest priority, followed by the configuration file, and finally the metadata server.
* **Updated `getMachineType` Function**: The `getMachineType` function has been modified to accept the `Config` object, enabling it to correctly evaluate the `machine-type` from the configuration file as part of the new precedence logic.
* **New Test Cases for Precedence**: Comprehensive unit and integration tests have been added to validate the new `machine-type` precedence order, covering scenarios where the machine type is set via CLI, config file, or metadata server.

### Link to the issue in case of a bug fix.
[b/464209951](http://b/464209951)

### Testing details
1. Manual - Verified by local run.
   - Before fix
     
     6SHgBQtbtGLD5T3-before-machine-type-fix

   - After fix
     
     
     57kjSiRFZ9uQNqz-after_machine_type_fix

3. Unit tests - NA
4. Integration tests - passed through presubmit ([log](http://fusion2/ci/kokoro/prod%3Agcsfuse%2Fgcp_ubuntu%2Fpresubmits%2Fperf_tests%2Fpresubmit/activity/c70f7a47-defe-423b-9ce6-53b53e7a01ed)) .

### CODE REVIEW

File: cfg/optimize_test.go [hunk 1]
```diff
@@ -130,12 +130,76 @@ func TestGetMachineType_FlagIsSet(t *testing.T) {
...
+	}
+	// Create a config where machine-type is set.
+	cfg := defaultConfig()
+	cfg.MachineType = "config-file-machine-type"
+
+	machineType, err := getMachineType(isSet, &cfg)
+
+	require.NoError(t, err)
+	assert.Equal(t, "cli-machine-type", machineType)
+}
```
[the remark this answers is not in the document]
Comment (gargnitingoogle) [reply cfg/optimize_test.go]: done

File: cmd/root_test.go [hunk 2]
```diff
@@ -924,6 +924,15 @@ func TestArgsParsing_FileSystemFlags(t *testing.T) {
...
 	expectedAIMLTrainingFileSystemConfig := expectedDefaultFileSystemConfig
+	createTempConfigFile := func(t *testing.T, content string) string {
+		t.Helper()
+		f, err := os.CreateTemp(t.TempDir(), "config.yaml")
+		require.NoError(t, err)
+		_, err = f.WriteString(content)
+		require.NoError(t, err)
+		require.NoError(t, f.Close())
+		return f.Name()
+	}
```
[the remark this answers is not in the document]
Comment (gargnitingoogle) [reply cmd/root_test.go]: done

File: cfg/optimize_test.go [hunk 3]
```diff
@@ -130,12 +130,75 @@ func TestGetMachineType_FlagIsSet(t *testing.T) {
...
+			metadataEndpoints = []string{server.URL}
+
+			var testConfig *Config
+			if tc.config != nil {
+				c := defaultConfig()
+				c.MachineType = tc.config.MachineType
+				testConfig = &c
+			}
+
+			machineType, err := getMachineType(tc.isSet, testConfig)
```
[the remark this answers is not in the document]
Comment (gargnitingoogle) [reply cfg/optimize_test.go]: done

File: cmd/config_validation_test.go [hunk 4]
```diff
@@ -127,6 +127,16 @@ func TestValidateConfigFile(t *testing.T) {
 			configFile: "testdata/valid_profile.yaml",
 			wantErr:    false,
 		},
+		{
+			name:       "set_machine_type",
```
Comment (gargnitingoogle) [inline cmd/config_validation_test.go]: Add this as a separate test which parses the config file to also confirm the machine-type value in the final config .

Comment (gargnitingoogle) [reply cmd/config_validation_test.go]: done

File: cfg/optimize.go [hunk 5]
```diff
@@ -103,14 +103,20 @@ func getMetadata(client *http.Client, endpoint string) ([]byte, error) {
...
+	// 1. Check if the machine-type flag is set in CLI flag.
 	if isSet.IsSet(machineTypeFlg) {
 		if currentMachineType := isSet.GetString(machineTypeFlg); currentMachineType != "" {
 			return currentMachineType, nil
 		}
 	}
+	// 2. Check if machine-type flag is set in config-file.
+	if cfg != nil && cfg.MachineType != "" {
+		return cfg.MachineType, nil
+	}
```
Comment (kislaykishore) [inline cfg/optimize.go:118]: Still unclear to me why the previous case, i.e. isSet.IsSet(machineType) doesn't handle this.

Comment (gargnitingoogle) [reply cfg/optimize.go:118]: yeah me too. This is exactly why I too didn't realize this issue initially. I'll investigate this, but as discussed, I won't block the fix on this investigation.

File: cmd/testdata/set_machine_type.yaml
Comment (ashmeenkaur) [inline cmd/testdata/set_machine_type.yaml:1]: can we re-use [this](https://github.com/GoogleCloudPlatform/gcsfuse/blob/master/cmd/testdata/valid_config.yaml) file instead?

File: cmd/root_test.go [hunk 6]
```diff
@@ -1154,6 +1172,24 @@ func TestArgsParsing_FileSystemFlags(t *testing.T) {
...
+		},
+		{
+			name: "config_file_overrides_metadata_server",
+			args: []string{"gcsfuse", "--config-file", createTempConfigFile(t, "machine-type: config-file-type"), "abc", "pqr"},
+			expectedConfig: &cfg.Config{
+				FileSystem:  expectedDefaultFileSystemConfig,
+				MachineType: "config-file-type",
+			},
+			checkMachineType: true,
+		},
```
Comment (ashmeenkaur) [inline cmd/root_test.go]: these tests are same as config_validation_test.go I think. Can we de-dupe them?

File: cfg/optimize_test.go [hunk 7]
```diff
@@ -130,12 +130,68 @@ func TestGetMachineType_FlagIsSet(t *testing.T) {
...
 
+func TestGetMachineType_InputPrecedenceOrder(t *testing.T) {
+	tests := []struct {
+		name                string
+		isSet               *mockIsValueSet
+		config              *Config
+		expectedMachineType string
+	}{
+		{
+			name: "CLI flag set",
```
Comment (ashmeenkaur) [inline cfg/optimize_test.go]: Nit: can we use '_' in the name instead of space for easy correlation/searchability?
https://g3doc.corp.google.com/company/teams/go-community/gotip/episodes/117.md?cl=head

File: cfg/optimize_test.go [hunk 8]
```diff
@@ -130,12 +130,68 @@ func TestGetMachineType_FlagIsSet(t *testing.T) {
...
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			resetMetadataEndpoints(t)
+			// Create a test server that returns a machine type.
+			server := createTestServer(t, func(w http.ResponseWriter, r *http.Request) {
+				fmt.Fprint(w, "zones/us-central1-a/machineTypes/n1-standard-1")
+			})
+			defer closeTestServer(t, server)
+			// Override metadataEndpoints for testing.
+			metadataEndpoints = []string{server.URL}
```
Comment (ashmeenkaur) [inline cfg/optimize_test.go:191]: I don't understand why this is required? IIUC, `getMachineType` function will return before reaching the metadata server in these tests.

File: cfg/optimize_test.go [hunk 9]
```diff
@@ -130,12 +130,68 @@ func TestGetMachineType_FlagIsSet(t *testing.T) {
...
+		config              *Config
+		expectedMachineType string
+	}{
+		{
+			name: "CLI flag set",
+			isSet: &mockIsValueSet{
+				setFlags:    map[string]bool{"machine-type": true},
+				stringFlags: map[string]string{"machine-type": "cli-machine-type"},
+			},
+			config:              nil,
```
Comment (ashmeenkaur) [inline cfg/optimize_test.go:152]: let's add test for empty config too

File: cfg/optimize.go (hunk 5 above)
Comment (ashmeenkaur) [reply cfg/optimize.go:118]: I think this is because we are using pflag's isSet in root.go: https://github.com/GoogleCloudPlatform/gcsfuse/blob/86f4cb1e74c6a06a68d56fdf31736920194c5030/cmd/root.go#L55

It is viper's isSet which tells us about all locations.

Comment (gargnitingoogle) [reply cfg/optimize.go:118]: Discussed this offline and resolved why isSet didn't work here.

File: cfg/optimize_test.go (hunk 9 above)
Comment (gargnitingoogle) [reply cfg/optimize_test.go:152]: added

File: cfg/optimize_test.go (hunk 7 above)
Comment (gargnitingoogle) [reply cfg/optimize_test.go]: done

File: cmd/testdata/set_machine_type.yaml
Comment (gargnitingoogle) [reply cmd/testdata/set_machine_type.yaml:1]: done

File: cfg/optimize_test.go (hunk 8 above)
Comment (gargnitingoogle) [reply cfg/optimize_test.go:191]: yes, but without this we can't confirm that the machine-type coming from metadata server will not override the value coming from cli flag or config-file. So this is needed.
